core/dispatcher/src/main/scala/whisk/core/invoker/Invoker.scala

```diff
@@ -311,27 +316,49 @@ class Invoker(
      *
      * Note: Updates the container's log cursor to indicate consumption of log.
      */
-    private def getContainerLogs(con: WhiskContainer, tries: Int = LogRetryCount)(implicit transid: TransactionId): JsArray = {
+    private def getContainerLogs(con: WhiskContainer, limits: ActionLimits, tries: Int = LogRetryCount)(implicit transid: TransactionId): JsArray = {
         val size = con.getLogSize(runningInContainer)
+        val maxSize = limits.logs.map { _.megabytes }.getOrElse(Int.MaxValue)
         val advanced = size != con.lastLogSize
         if (tries <= 0 || advanced) {
             val rawLogBytes = con.getDockerLogContent(con.lastLogSize, size, runningInContainer)
             val rawLog = new String(rawLogBytes, "UTF-8")
+
             val isNodeJs = con.image == nodejsImageName || con.image == nodejs6ImageName
-            var isSwift = con.image == swiftImageName || con.image == swift3ImageName
-            val (complete, logArray) = processJsonDriverLogContents(rawLog, isNodeJs || isSwift)
+            val isSwift = con.image == swiftImageName || con.image == swift3ImageName
+            val (complete, logArray) = processJsonDriverLogContents(rawLog, isNodeJs || isSwift, maxSize megabytes)
+
             if (tries > 0 && !complete) {
                 info(this, s"log cursor advanced but missing sentinel, trying $tries more times")
                 Thread.sleep(LogRetry)
-                getContainerLogs(con, tries - 1)
+                getContainerLogs(con, limits, tries - 1)
             } else {
                 con.lastLogSize = size
                 logArray
             }
         } else {
             info(this, s"log cursor has not advanced, trying $tries more times")
             Thread.sleep(LogRetry)
-            getContainerLogs(con, tries - 1)
+            getContainerLogs(con, limits, tries - 1)
+        }
+    }
+
+    /**
+     * Takes the first bytes of a list of strings. If the number of bytes would cut an entry
+     * it is omitted entirely.
+     *
+     * @param strings The list of strings
+     * @param bytes number of
+     */
+    def truncateLogs(strings: Vector[String], sizeLimit: ObjectSize, written: ObjectSize = 0 bytes): Vector[String] = {
+        val stringStream = strings.toStream
+        val cummulativeSizes = stringStream.scanLeft(0.bytes) { (acc, current) => acc + current.sizeInBytes }.tail
+        val truncatedLogs = stringStream.zip(cummulativeSizes).takeWhile(_._2 < sizeLimit).map(_._1).toVector
```

Comment:
It's functional but takes 5 passes on something that's potentially "large". What about one scanLeft and make the accumulator the vector and the cumulative size.